### PR TITLE
libtiff: add missing dependencies

### PR DESCRIPTION
**Pull request by glennpj**
- gl
- glu
- freeglut

### CODE REVIEW

**Comment by adamjstewart** [comment]
It looks like GL and friends are only needed to build `tiffgt`:
```console
$ otool -L /Users/Adam/spack/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/libtiff-4.3.0-3ud2nw6m4snhlfvcdes7wo7i2jcw4nwx/bin/tiffgt 
/Users/Adam/spack/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/libtiff-4.3.0-3ud2nw6m4snhlfvcdes7wo7i2jcw4nwx/bin/tiffgt:
	/Users/Adam/spack/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/libtiff-4.3.0-3ud2nw6m4snhlfvcdes7wo7i2jcw4nwx/lib/libtiff.5.dylib (compatibility version 13.0.0, current version 13.0.0)
	/opt/X11/lib/libglut.3.dylib (compatibility version 13.0.0, current version 13.0.0)
	/opt/X11/lib/libSM.6.dylib (compatibility version 7.0.0, current version 7.1.0)
	/opt/X11/lib/libICE.6.dylib (compatibility version 10.0.0, current version 10.0.0)
	/opt/X11/lib/libXi.6.dylib (compatibility version 8.0.0, current version 8.0.0)
	/opt/X11/lib/libGLU.1.dylib (compatibility version 5.0.0, current version 5.1.0)
	/opt/X11/lib/libGL.1.dylib (compatibility version 4.0.0, current version 4.0.0)
	/opt/X11/lib/libX11.6.dylib (compatibility version 11.0.0, current version 11.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
	@rpath/liblzma.5.dylib (compatibility version 8.0.0, current version 8.5.0)
	@rpath/libjpeg.62.dylib (compatibility version 62.0.0, current version 62.3.0)
	@rpath/libz.1.2.11.dylib (compatibility version 1.0.0, current version 1.2.11)
```
So we can either add these 7 dependencies or we can add `--without-x` which seems to disable the build of `libgt` altogether. We can also make it a variant if you want.

**Comment by glennpj** [comment]
An X variant might make sense. I just noticed that libzstd is getting pulled in from the system even though the `zstd` variant is off and `configure` was passed `--disable-zstd`. I have some other variants enabled so something must be turning that on. I will have to track that down as well.

**Comment by glennpj** [comment]
This is actually related to #28185. That PR adds `force_autoreconf` and when `autoreconf` is run the configure script is radically different, making OpenGL non-optional. Here is the tail end of configure pre and post autoreconf:

#### pre-autoreconf
```console
Some influential environment variables:
  CC          C compiler command
  CFLAGS      C compiler flags
  LDFLAGS     linker flags, e.g. -L<lib dir> if you have libraries in a
              nonstandard directory <lib dir>
  LIBS        libraries to pass to the linker, e.g. -l<library>
  CPPFLAGS    (Objective) C/C++ preprocessor flags, e.g. -I<include dir> if
              you have headers in a nonstandard directory <include dir>
  LT_SYS_LIBRARY_PATH
              User-defined run-time library search path.
  CPP         C preprocessor
  CXX         C++ compiler command
  CXXFLAGS    C++ compiler flags
  CXXCPP      C++ preprocessor
  XMKMF       Path to xmkmf, Makefile generator for X Window System

Use these variables to override the choices made by `configure' or to help
it to find libraries and programs with nonstandard names/locations.

Report bugs to <tiff@lists.maptools.org>.
```

#### post-autoreconf
```console
Some influential environment variables:
  CC          C compiler command
  CFLAGS      C compiler flags
  LDFLAGS     linker flags, e.g. -L<lib dir> if you have libraries in a
              nonstandard directory <lib dir>
  LIBS        libraries to pass to the linker, e.g. -l<library>
  CPPFLAGS    (Objective) C/C++ preprocessor flags, e.g. -I<include dir> if
              you have headers in a nonstandard directory <include dir>
  LT_SYS_LIBRARY_PATH
              User-defined run-time library search path.
  CPP         C preprocessor
  CXX         C++ compiler command
  CXXFLAGS    C++ compiler flags
  CXXCPP      C++ preprocessor
  XMKMF       Path to xmkmf, Makefile generator for X Window System
  PKG_CONFIG  path to pkg-config utility
  PKG_CONFIG_PATH
              directories to add to pkg-config's search path
  PKG_CONFIG_LIBDIR
              path overriding pkg-config's built-in search path
  GL_CFLAGS   C compiler flags for GL, overriding configure script defaults
  GL_LIBS     Linker flags for GL, overriding configure script defaults
  GLU_CFLAGS  C compiler flags for GLU, overriding system check
  GLU_LIBS    Linker flags for GLU, overriding system check
  GLUT_CFLAGS C compiler flags for GLUT, overriding configure script defaults
  GLUT_LIBS   Linker flags for GLUT, overriding configure script defaults

Use these variables to override the choices made by `configure' or to help
it to find libraries and programs with nonstandard names/locations.

Report bugs to <tiff@lists.maptools.org>.
```

If I back out #28185 then `libtiff` does not need OpenGL, or X, and `tiffgt` is not built. It seems that this is an unexpected consequence of #28185 so I was wondering how to proceed. Should #28185 be Mac only?

If we leave the package as-is, then an X variant does not help because the configure script will still look for OpenGL. If it is found, libtiff builds, if it is not found then it fails at the configure phase. More patches would be needed to make this work.

**Comment by adamjstewart** [comment]
That is quite frustrating. I think I originally tried directly patching `configure` instead of `configure.ac` but that didn't work for some reason and the autoreconf solution was needed. It looks like macports came up with a completely different hack: https://trac.macports.org/ticket/62758. I wish the developers would release a new version of libtiff so we could just remove the patch and ignore that problem. 

I'm personally fine with only applying that patch to macOS, or only to Apple Clang. Based on https://gitlab.com/libtiff/libtiff/-/issues/252 it seems like this bug affects all case-insensitive filesystems (macOS by default, but can be changed) when using a C++20 compiler. I don't really know how to properly check for this.

**Comment by glennpj** [comment]
I am not at all familiar with using Spack on Macs. Would the GL dependencies that I added be needed for that case or can it just be assumed that Macs have OpenGL already?

**Comment by adamjstewart** [comment]
It looks like all 7 of those deps would be needed, we don't want to link to system libs and GL and friends may only be present when X11 is installed on the system. Once a new libtiff release is added we can remove all of this.

File: var/spack/repos/builtin/packages/libtiff/package.py [hunk 1]
```diff
@@ -56,16 +56,27 @@ class Libtiff(AutotoolsPackage):
...
+    depends_on('libtool', when='@4.3.0 platform=darwin', type='build')
+    depends_on('m4', when='@4.3.0 platform=darwin', type='build')
 
     @property
     def force_autoreconf(self):
-        return self.spec.satisfies('@4.3.0')
+        if self.spec.satisfies('platform=darwin'):
+            return self.spec.satisfies('@4.3.0')
+        else:
+            return False
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/libtiff/package.py]: How is this different than:
```suggestion
        return self.spec.satisfies('@4.3.0 platform=darwin')
```
?

Comment (glennpj) [reply var/spack/repos/builtin/packages/libtiff/package.py]: Ha, it is not different, my brain just went in a different direction.

**Review by adamjstewart — APPROVED** [review]
Looks good now. I'll remove all this cruft once a new libtiff release comes out.

**Comment by sethrj** [comment]
Yuck, this ends up by default causing llvm to install on my mac :(

**Comment by sethrj** [comment]
@glennpj @adamjstewart can we change "preferred" to 4.2.0? This ends up trying to build a whole X11 stack on macos.

**Comment by adamjstewart** [comment]
Looks like there is a 4.4.0rc1 version out now. I might just drop the macOS patch/deps and anyone who has issues can use that version instead.

**Pull request merged**